cosmos/x/evm/plugins/txpool/mempool/mempool_writer.go

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: BUSL-1.1
+//
+// Copyright (C) 2023, Berachain Foundation. All rights reserved.
+// Use of this software is govered by the Business Source License included
+// in the LICENSE file of this repository and at www.mariadb.com/bsl11.
+//
+// ANY USE OF THE LICENSED WORK IN VIOLATION OF THIS LICENSE WILL AUTOMATICALLY
+// TERMINATE YOUR RIGHTS UNDER THIS LICENSE FOR THE CURRENT AND ALL OTHER
+// VERSIONS OF THE LICENSED WORK.
+//
+// THIS LICENSE DOES NOT GRANT YOU ANY RIGHT IN ANY TRADEMARK OR LOGO OF
+// LICENSOR OR ITS AFFILIATES (PROVIDED THAT YOU MAY USE A TRADEMARK OR LOGO OF
+// LICENSOR AS EXPRESSLY REQUIRED BY THIS LICENSE).
+//
+// TO THE EXTENT PERMITTED BY APPLICABLE LAW, THE LICENSED WORK IS PROVIDED ON
+// AN “AS IS” BASIS. LICENSOR HEREBY DISCLAIMS ALL WARRANTIES AND CONDITIONS,
+// EXPRESS OR IMPLIED, INCLUDING (WITHOUT LIMITATION) WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT, AND
+// TITLE.
+
+package mempool
+
+import (
+	"context"
+	"errors"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	evmtypes "pkg.berachain.dev/polaris/cosmos/x/evm/types"
+	"pkg.berachain.dev/polaris/eth/common"
+)
+
+// SetNonceRetriever sets the nonce retriever db for the mempool.
+func (etp *EthTxPool) SetNonceRetriever(nr NonceRetriever) {
+	etp.nr = nr
+}
+
+// Insert is called when a transaction is added to the mempool.
+func (etp *EthTxPool) Insert(ctx context.Context, tx sdk.Tx) error {
+	etp.mu.Lock()
+	defer etp.mu.Unlock()
+
+	// Reject txs with a nonce lower than the nonce reported by the statedb.
+	if sdbNonce := etp.nr.GetNonce(
+		common.BytesToAddress(tx.GetMsgs()[0].GetSigners()[0]),
```

Comment:
can make this a helper func, getTxSigner/Sender

Reply:
👍